programs/stake/src/stake_state.rs

```diff
@@ -6431,6 +6432,374 @@ mod tests {
+        let starting_stake_delegation = MINIMUM_STAKE_DELEGATION;
+        for (ending_stake_delegation, expected_result) in [
+            (MINIMUM_STAKE_DELEGATION, Ok(())),
+            (
+                MINIMUM_STAKE_DELEGATION - 1,
+                Err(InstructionError::InsufficientFunds),
+            ),
+        ] {
+            let rent = Rent::default();
+            let rent_exempt_reserve = rent.minimum_balance(std::mem::size_of::<StakeState>());
+            let stake_pubkey = Pubkey::new_unique();
+            let meta = Meta {
+                rent_exempt_reserve,
+                ..Meta::auto(&stake_pubkey)
+            };
+
+            for stake_state in &[
+                StakeState::Initialized(meta),
+                StakeState::Stake(meta, just_stake(starting_stake_delegation)),
+            ] {
+                let rewards_balance = 123;
+                let stake_account = AccountSharedData::new_ref_data_with_space(
+                    starting_stake_delegation + rent_exempt_reserve + rewards_balance,
+                    stake_state,
+                    std::mem::size_of::<StakeState>(),
+                    &id(),
+                )
+                .unwrap();
+                let stake_keyed_account = KeyedAccount::new(&stake_pubkey, true, &stake_account);
+
+                let to_pubkey = Pubkey::new_unique();
+                let to_account =
+                    AccountSharedData::new_ref(rent_exempt_reserve, 0, &system_program::id());
+                let to_keyed_account = KeyedAccount::new(&to_pubkey, false, &to_account);
+
+                let withdraw_amount =
+                    (starting_stake_delegation + rewards_balance) - ending_stake_delegation;
+                assert_eq!(
+                    expected_result,
+                    stake_keyed_account.withdraw(
+                        withdraw_amount,
+                        &to_keyed_account,
+                        &Clock::default(),
+                        &StakeHistory::default(),
+                        &stake_keyed_account,
+                        None,
+                    ),
+                );
+            }
+        }
+    }
+
+    /// The stake program currently allows delegations below the minimum stake delegation (see also
+    /// `test_delegate_minimum_stake_delegation()`).  This is not the ultimate desired behavior,
+    /// but this test ensures the existing behavior is not changed inadvertently.
+    ///
+    /// This test:
+    /// 1. Initialises a stake account (with sufficient balance for both rent and minimum delegation)
+    /// 2. Delegates the minimum amount
+    /// 3. Deactives the delegation
+    /// 4. Withdraws from the account such that the ending balance is *below* rent + minimum delegation
+    /// 5. Re-delegates, now with less than the minimum delegation, but it still succeeds
+    #[test]
+    fn test_behavior_withdrawl_then_redelegate_with_less_than_minimum_stake_delegation() {
```

Comment:
@ryoqun I've added this test to address your comment from here: https://github.com/solana-labs/solana/pull/22663#discussion_r805955838

Reply:
```suggestion
    /// The stake program currently allows delegations below the minimum stake delegation (see also
    /// `test_delegate_minimum_stake_delegation()`).  This is not the ultimate desired behavior,
    /// but this test ensures the existing behavior is not changed inadvertently.
    ///
    /// This test:
    /// 1. Initialises a stake account (with sufficient balance for both rent and minimum delegation)
    /// 2. Delegates the minimum amount
    /// 3. Deactives the delegation
    /// 4. Withdraws from the account such that the ending balance is *below* rent + minimum delegation
    /// 5. Re-delegates, now with less than the minimum delegation, but it still succeeds
    #[test]
    fn test_behavior_withdrawal_then_redelegate_with_less_than_minimum_stake_delegation() {
```